File: modules/grove-volume/android/src/main/java/expo/modules/grovevolume/GroveVolumeModule.kt
package expo.modules.grovevolume

import android.content.Context
import android.media.AudioManager
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class GroveVolumeModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("GroveVolume")
    AsyncFunction("getOutputVolume") {
      val manager = appContext.reactContext?.getSystemService(Context.AUDIO_SERVICE) as? AudioManager
      if (manager == null) null else {
        val maximum = manager.getStreamMaxVolume(AudioManager.STREAM_MUSIC)
        if (maximum <= 0) null else manager.getStreamVolume(AudioManager.STREAM_MUSIC).toDouble() / maximum
      }
    }
  }
}

File: modules/grove-volume/ios/GroveVolumeModule.swift
import ExpoModulesCore
import AVFoundation

public class GroveVolumeModule: Module {
  public func definition() -> ModuleDefinition {
    Name("GroveVolume")
    AsyncFunction("getOutputVolume") { () -> Double in
      return Double(AVAudioSession.sharedInstance().outputVolume)
    }
  }
}
